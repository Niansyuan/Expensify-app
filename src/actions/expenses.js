import { child, get, push, ref, remove, update } from '@firebase/database';
import database from '../firebase/firebase';

//ADD_EXPENSES
// const action=(({}={})=>({}))
export const addExpenses = (expense) => ({
    type: 'ADD_EXPENSES',
    expense
});

//dispatch addExpenses to change store (add this action after introduce firebase)
export const startAddExpenses = (expenseDate = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createAt = 0
        } = expenseDate;
        const expense = { description, note, amount, createAt }
        const rootRef = ref(database, `users/${uid}/expenses`)
        return push(rootRef, expense).then((rootRef) => {
            dispatch(addExpenses({
                id: rootRef.key,
                ...expense
            }));
        });
    };
};

//REMOVE_EXPENSES
export const removeExpenses = ({ id }) => ({
    type: 'REMOVE_EXPENSES',
    id
})

export const startRemoveExpenses = ({ id }) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return remove(ref(database, `users/${uid}/expenses/${id}`)).then(() => {
            dispatch(removeExpenses({ id }))
        });
    };
};

//EDIT_EXPENSES
export const editExpenses = (id, updates) => ({
    type: 'EDIT_EXPENSES',
    id,
    updates
});

export const startEditExpenses = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return update(ref(database, `users/${uid}/expenses/${id}`), updates).then(() => {
            dispatch(editExpenses(id, updates))
        });
    };
};


//SET_EXPENSES (add this action after introduce firebase)
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//startSetExpenses (add this action after introduce firebase)
export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        //fetch all expenseData once
        return get(child(ref(database), `users/${uid}/expenses`)).then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            });
            dispatch(setExpenses(expenses));
        });
    };
};