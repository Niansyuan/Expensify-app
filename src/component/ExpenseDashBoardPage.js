import React from 'react'

const ExpenseDashBoardPage=()=>(
    <div>
        <p>This is from dashBoard component</p>
    </div>
)

export default ExpenseDashBoardPage