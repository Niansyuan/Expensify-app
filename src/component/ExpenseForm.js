import "react-dates/initialize";
import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
    state={
        description:'',
        note:'',
        amount:'',
        createAt:moment(),
        calendarFocused: false,
        error:''
    };
    handleDiscriptionChange=(e)=>{
        const description=e.target.value;
        this.setState(()=>({description}));
    }
    handleNoteChange=(e)=>{
        const note=e.target.value;
        this.setState(()=>({note}));
    }
    handleAmountChange=(e)=>{
        const amount=e.target.value;
        if(amount.match(/^\d{1,}(\.\d{0,2})?$/)){ //regular expression,詳看notion筆記
            this.setState(()=>({amount}))
        }
    }
    handleDateChange=(createAt)=>{
        this.setState(()=>({createAt}))
    }
    handleFocusChange=({focused})=>{
        this.setState(()=>({calendarFocused:focused}))
    }
    handleSubmit=(e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error:'Please Provide Description and Amount'}))
        }else{
            this.setState(()=>({error:''}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10),
                createAt:this.state.createAt.valueOf(),
                note:this.state.note
            })
        }
    }
    render(){
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.handleSubmit}>
                    <input 
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.handleDiscriptionChange}
                    />
                    <input
                    type="number"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.handleAmountChange}
                    />
                    <SingleDatePicker
                    date={this.state.createAt}
                    onDateChange={this.handleDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.handleFocusChange}
                    numberOfMonths={1} //一次顯示幾個月
                    isOutsideRange={()=>false} 
                    />
                    <textarea
                    placeholder="Add a note for your expense...(optional)"
                    value={this.state.note}
                    onChange={this.handleNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}