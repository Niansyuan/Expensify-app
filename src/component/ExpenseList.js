import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpense from "../selectors/expenses";

export const ExpenseList=(props)=>(
    <div>
        {
            props.expenses.length ===0 ? (
                <p>No expenses</p>
            ):(
                props.expenses.map((expense)=>{
                    return <ExpenseListItem key={expense.id} {...expense}/>;
                })
            )
        }
    </div>
)

//HOC -------
const mapStateToProps=(state)=>{
    return {
        expenses:selectExpense(state.expenses,state.filters)
    }
}
export default connect(mapStateToProps)(ExpenseList);
//HOC -------