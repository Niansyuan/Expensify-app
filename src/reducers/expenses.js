//expenses reducer
const expensesReducerDefaultState=[]
const expensesReducer=(state=expensesReducerDefaultState,action)=>{
    switch(action.type){
        case 'ADD_EXPENSES':
            return [...state,action.expense]; //等於state.concat(action.expense)
        case 'REMOVE_EXPENSES':
            return state.filter(({id})=>{
                return id!== action.id
            });
        case 'EDIT_EXPENSES':
            return state.map((expense)=>{
                if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                }else{
                    return expense
                }
            });
        default:
            return state;
    }
}
export default expensesReducer;