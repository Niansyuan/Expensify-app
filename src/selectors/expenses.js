import moment from "moment";

//Get visible expenses
export default(expenses,{text,sortBy,startDate,endDate})=>{
    return expenses.filter((expense)=>{
        //all three are ture,we want to return true from filter
        const createAtMoment=moment(expense.createAt);
        const startDateMatch= startDate ? startDate.isSameOrBefore(createAtMoment,'day'):true;
        const endDateMatch= endDate ? endDate.isSameOrAfter(createAtMoment,'day'):true;
        const textMatch= expense.description.toLowerCase().includes((text.toLowerCase()));
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b)=>{
        if(sortBy === 'date'){
            return a.createAt < b.createAt? 1:-1
        };
        if(sortBy === 'amount'){
            return a.amount < b.amount? 1:-1
        };
    })
}