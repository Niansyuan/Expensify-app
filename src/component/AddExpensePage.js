import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { addExpenses } from '../actions/expenses'

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpenses(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpenses: (expense) => dispatch(addExpenses(expense))
    });

export default connect(undefined, mapDispatchToProps)(AddExpensePage);