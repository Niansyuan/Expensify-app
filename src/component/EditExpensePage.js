import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm'
import { editExpenses, removeExpenses } from '../actions/expenses'

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpenses(this.props.expense.id, expense)
        this.props.history.push('/')
        console.log('update', expense)
    }
    onClick = () => {
        this.props.removeExpenses({ id: this.props.expense.id })
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button
                    expense={this.props.expense}
                    onClick={this.onClick}>Remove</button>
            </div>
        )

    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({
    editExpenses:(id,expense)=>dispatch(editExpenses(id,expense)),
    removeExpenses:(data)=>dispatch(removeExpenses(data))
})

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);