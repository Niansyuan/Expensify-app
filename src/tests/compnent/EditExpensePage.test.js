import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../component/EditExpensePage";
import expenses from '../fixture/expenses';
import '../setupTests';

let editExpenses,removeExpenses,history,wrapper;

beforeEach(()=>{
    editExpenses=jest.fn();
    removeExpenses=jest.fn();
    history={push:jest.fn()};
    wrapper=shallow(<EditExpensePage
        editExpenses={editExpenses} 
        removeExpenses={removeExpenses} 
        history={history}
        expense={expenses[1]}
    />)
})

test('should render EditExpensePage',()=>{
    expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense',()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
    expect(editExpenses).toHaveBeenLastCalledWith(expenses[1].id,expenses[1])
    expect(history.push).toHaveBeenLastCalledWith('/')
});

test('should render removeExpense',()=>{
    wrapper.find('button').simulate('click')
    expect(removeExpenses).toHaveBeenLastCalledWith({id:expenses[1].id})
    expect(history.push).toHaveBeenLastCalledWith('/')
});