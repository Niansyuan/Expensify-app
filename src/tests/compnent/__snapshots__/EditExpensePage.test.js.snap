// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 8500,
        "createAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    expense={
      Object {
        "amount": 8500,
        "createAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
